### PR TITLE
Add retry logic to check encode quality

### PR DESCRIPTION
**Pull request by e-dong**
Addresses #31 

## Summary of Changes
* Added timeout and listen_timeout to listen url [FFmpeg Protocols Documentation#tcp](https://ffmpeg.org/ffmpeg-protocols.html#tcp)
* Added retry logic when vmaf child process times out
* Ran `rustfmt` to format edited files  
   
2023-09-26-215459_1920x436_scrot

### CODE REVIEW

**Comment by e-dong** [comment]
@Proryanator Standby, gonna rebase this branch. The commit history is pretty messy lol

**Review by Proryanator — CHANGES_REQUESTED** [review]
This is great! :D no more getting stuck when trying to calculate the vmaf score.

A few comments but otherwise looks great!

Also also, I'm going to add an issue to update the README with how to run rustfmt, and how to run that within your IDE as well (huge fan of standardized formatting when collaborating).

File: engine/src/permutation_engine.rs [hunk 1]
```diff
@@ -65,7 +71,31 @@ impl PermutationEngine {
 
             if !result.was_overloaded && permutation.check_quality.clone() {
                 let vmaf_start_time = SystemTime::now();
-                result.vmaf_score = check_encode_quality(permutation.clone(), &ctrl_channel, permutation.verbose);
+                let mut quality_val = None;
+                for attempt in 0..MAX_ATTEMPTS_CALC_QUALITY {
+                    print!("  [ATTEMPT {}/{}] ", attempt + 1, MAX_ATTEMPTS_CALC_QUALITY);
```
[the remark this answers is not in the document]
Comment (e-dong) [reply engine/src/permutation_engine.rs:76]: Yeah absolutely, the end user probably wouldn't care haha.

File: engine/src/permutation_engine.rs [hunk 2]
```diff
@@ -65,7 +71,31 @@ impl PermutationEngine {
 
             if !result.was_overloaded && permutation.check_quality.clone() {
                 let vmaf_start_time = SystemTime::now();
-                result.vmaf_score = check_encode_quality(permutation.clone(), &ctrl_channel, permutation.verbose);
```
[the remark this answers is not in the document]
Comment (e-dong) [reply engine/src/permutation_engine.rs:68]: Yeah that makes sense!

File: engine/src/permutation_engine.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (e-dong) [reply engine/src/permutation_engine.rs:76]: 2023-09-28-203146_1388x128_scrot

Yeah that's a lot cleaner!

[the remark this answers is not in the document]
Comment (e-dong) [reply engine/src/permutation_engine.rs]: Added some more verbose logs
```
[00:00:10] [----------------] 0/2383 frames (00:00:00)                                                      VMAF calculation finishing up...                       tcp://localhost:2000?listen&listen_timeout=3000&timeout=1000000: Connection timed out
See perm-278-attempt-1-ffmpeg-20230929-033554.log for more details.                                           Check encode quality failed. Retrying...               [ ATTEMPT 2/3 ] Calculating vmaf score; might take longer than original encode depending on your CPU...
...
```

[the remark this answers is not in the document]
Comment (e-dong) [reply engine/src/permutation_engine.rs]: I actually might swap the filename around, so the attempts are sorted by timestamp first.
Failed 3 times.

```
perm-175-attempt-1-ffmpeg-20230929-022024.log
      perm-253-attempt-1-ffmpeg-20230929-031736.log
      perm-278-attempt-1-ffmpeg-20230929-033554.log
```

**Review by Proryanator — APPROVED** [review]

**Pull request merged**